src/ccs/multifolding.rs

```diff
@@ -56,29 +58,51 @@ impl<G: Group> NIMFS<G> {
       ccs_mle,
       ck,
       lcccs,
+      transcript: TranscriptEngineTrait::new(b"NIMFS"),
     }
   }
 
   /// Initializes a NIMFS instance given the CCS of it and a first witness vector that satifies it.
   // XXX: This should probably return an error as we should check whether is satisfied or not.
-  pub fn init<R: RngCore>(mut rng: &mut R, ccs: CCS<G>, z: Vec<G::Scalar>) -> Self {
+  pub fn init(ccs: CCS<G>, z: Vec<G::Scalar>, seed: &'static [u8]) -> Self {
+    let mut transcript: G::TE = TranscriptEngineTrait::new(seed);
     let ccs_mle: Vec<MultilinearPolynomial<G::Scalar>> =
       ccs.M.iter().map(|matrix| matrix.to_mle()).collect();
+    // Add ccs circuit description to transcript. XXX: This does not need to be kept.
+    TranscriptEngineTrait::<G>::absorb(
+      &mut transcript,
+      b"ccs_matrixes",
+      &ccs_mle
+        .iter()
+        .flat_map(|mle| mle.Z.clone())
+        .collect::<Vec<G::Scalar>>(),
+    );
+
+    // Add the first round of witness to the transcript.
     let w: Vec<G::Scalar> = z[(1 + ccs.l)..].to_vec();
+    TranscriptEngineTrait::<G>::absorb(&mut transcript, b"og_w", &w);
+
     let ck = ccs.commitment_key();
-    let r_w = G::Scalar::random(&mut rng);
+    // XXX: Pedersen commitment API doesn't give a way to handle this.
+    // let r_w = G::Scalar::random(&mut rng);
     let w_comm = <G as Group>::CE::commit(&ck, &w);
 
-    let r_x: Vec<G::Scalar> = vec![G::Scalar::random(&mut rng); ccs.s];
-    let v = ccs.compute_v_j(&z, &r_x, &ccs_mle);
+    // Query challenge to get initial `r_x`.
```

Comment:
Yikes, 4 lines to squeeze a challenge? Can we add helper functions for this? Maybe just make `gen_r_x()` take as input the size?